syntax = "proto3";

package memory_service;

// Memory Service Definition
service MemoryService {
  rpc StoreMemory(StoreMemoryRequest) returns (StoreMemoryResponse);
  rpc GetMemory(GetMemoryRequest) returns (GetMemoryResponse);
  rpc UpdateMemory(UpdateMemoryRequest) returns (UpdateMemoryResponse);
  rpc ClearMemory(ClearMemoryRequest) returns (ClearMemoryResponse);
  rpc ListMemories(ListMemoriesRequest) returns (ListMemoriesResponse);
}

// Store Memory Request/Response
message StoreMemoryRequest {
  string user_id = 1;
  string tenant_id = 2;
  string key = 3;
  string value = 4;  // JSON string
  int32 ttl = 5;     // TTL in seconds (default 3600)
}

message StoreMemoryResponse {
  bool success = 1;
  string message = 2;
}

// Get Memory Request/Response
message GetMemoryRequest {
  string user_id = 1;
  string tenant_id = 2;
  string key = 3;
}

message GetMemoryResponse {
  bool found = 1;
  string value = 2;  // JSON string
  string message = 3;
}

// Update Memory Request/Response
message UpdateMemoryRequest {
  string user_id = 1;
  string tenant_id = 2;
  string key = 3;
  string value = 4;  // JSON string
}

message UpdateMemoryResponse {
  bool success = 1;
  string message = 2;
}

// Clear Memory Request/Response
message ClearMemoryRequest {
  string user_id = 1;
  string tenant_id = 2;
  string key = 3;     // Optional: if empty, clear all user memories
}

message ClearMemoryResponse {
  bool success = 1;
  string message = 2;
}

// List Memories Request/Response
message ListMemoriesRequest {
  string user_id = 1;
  string tenant_id = 2;
}

message MemoryItem {
  string key = 1;
  string value = 2;     // JSON string
  string created_at = 3;
  string expires_at = 4;
}

message ListMemoriesResponse {
  repeated MemoryItem memories = 1;
  int32 count = 2;
  string message = 3;
}